apis/v1alpha1/destination.go

```diff
@@ -18,132 +18,60 @@ package v1alpha1
-	// ObjectReference points to an Addressable.
-	*corev1.ObjectReference `json:",inline"`
+	// Ref points to an Addressable.
+	// +optional
+	Ref *corev1.ObjectReference `json:"ref,omitempty"`
 
-	// URI is for direct URI Designations.
+	// URI can be a absolute URI which points to the URI designation or it can be a relative URI will be merged to the URI retrieved from Ref.
+	// +optional
 	URI *apis.URL `json:"uri,omitempty"`
-
-	// Path is used with the resulting URL from Addressable ObjectReference or URI. Must start
-	// with `/`. An empty path should be represented as the nil value, not `` or `/`.  Will be
-	// appended to the path of the resulting URL from the Addressable, or URI.
-	Path *string `json:"path,omitempty"`
 }
 
 // NewDestination constructs a Destination from an object reference as a convenience.
-func NewDestination(obj *corev1.ObjectReference, paths ...string) (*Destination, error) {
-	dest := &Destination{
-		ObjectReference: obj,
-	}
-	err := dest.AppendPath(paths...)
-	if err != nil {
-		return nil, err
-	}
-	return dest, nil
+func NewDestination(obj *corev1.ObjectReference) (*Destination, error) {
+	return &Destination{
+		Ref: obj,
+	}, nil
 }
 
 // NewDestinationURI constructs a Destination from a URI.
-func NewDestinationURI(uri *apis.URL, paths ...string) (*Destination, error) {
-	dest := &Destination{
+func NewDestinationURI(uri *apis.URL) (*Destination, error) {
+	return &Destination{
 		URI: uri,
-	}
-	err := dest.AppendPath(paths...)
-	if err != nil {
-		return nil, err
-	}
-	return dest, nil
+	}, nil
 }
 
-// AppendPath iteratively appends paths to the Destination.
-// The path will always begin with "/" unless it is empty.
-// An empty path ("" or "/") will always resolve to nil.
-func (current *Destination) AppendPath(paths ...string) error {
-	// Start with empty string or existing path
-	var fullpath string
-	if current.Path != nil {
-		fullpath = *current.Path
-	}
-
-	// Intelligently join all the paths provided
-	fullpath = path.Join("/", fullpath, path.Join(paths...))
-
-	// Parse the URL to trim garbage
-	urlpath, err := apis.ParseURL(fullpath)
-	if err != nil {
-		return err
-	}
-
-	// apis.ParseURL returns nil if our path was empty, then our path
-	// should reflect that it is not set.
-	if urlpath == nil {
-		current.Path = nil
-		return nil
-	}
-
-	// A path of "/" adds no information, just toss it
-	// Note that urlpath.Path == "" is always false here (joined with "/" above).
-	if urlpath.Path == "/" {
-		current.Path = nil
-		return nil
-	}
-
-	// Only use the plain path from the URL
-	current.Path = &urlpath.Path
-	return nil
-}
 
 func (current *Destination) Validate(ctx context.Context) *apis.FieldError {
 	if current != nil {
-		errs := validateDestination(*current).ViaField(apis.CurrentField)
-		if current.Path != nil {
-			errs = errs.Also(validateDestinationPath(*current.Path).ViaField("path"))
-		}
-		return errs
+		return validateDestination(*current).ViaField(apis.CurrentField)
```

Comment:
`Validate(ctx context.Context)` implements  `Validatable` interface and receives `context.Context` as parameter. `ValidateDestination(dest Destination)` receives `Destination` as parameter. I found that in eventing, we will use this method `ValidateDestination(dest Destination)` to validate the subfields of `Trigger` in Trigger validation. We can't use `Validate(ctx context.Context)` to to validate the subfields of `Trigger` since the parameter doesn't match

> Why do we need `Validate` and `validateDestination`? Seems like these methods could be collapsed into just `Validate`. Any context @n3wscott?